# Optimization: Profit, Revenue, and Cost

Business and economics provide good examples of optimization problems - a company typically looks to maximize profits and minimize costs. In this example, we will use calculus with SymPy to maximize profit.

## Definitions

The **demand equation** for an item relates the price $p$ charged per item and the quantity $q$ of items that can be sold at that price (that is, $q$ can be thought of the number of people who are willing to buy the item at a price of $p$).

The **revenue** that a company generates from selling $q$ items at a price of $p$ is then $R = pq$.

If $C$ represents the **cost** of producing $q$ items, then the **profit** the company makes from selling $q$ items is the revenue minus the cost: $$P(q) = R(q) - C(q)$$

Before working on the problem, import in order to use SymPy.

In [1]:
from sympy import *

## Example

### Finding a Cost Function
Production of an item has fixed costs of $\$10,000$ and variable costs of $\$2$ per item. Express the cost, $C$, of producing $q$ items.

---

The variable costs for $q$ items are $2q$ dollars. To find the total cost of producing $q$ items, we add the fixed cost of $10000$ to the variable cost: $$C(q) = 2q + 10000$$

We can store the cost function $C(q)$ as C in Python, after defining q as a symbolic variable.

In [2]:
q = symbols('q')
C = 2*q + 10000

### Finding a Demand Equation

The relationship between the price, $p$, and quantity, $q$, demanded is linear. Market research shows that $10,100$ items are sold when the price is $\$5$ and $12,872$ items are sold when the price is $\$4.50$. Express the price $p$ as a function of quantity $q$.

---

Based on the given information, $p(q)$ will be a linear function.

We are told two coordinate points on the graph of the linear function $q(p)$.

"$10,100$ items are sold when the price is $\$5$," means the point $(q, p) = (10100, 5)$ lies on the graph.

"$12,872$ items are sold when the price is $\$4.50$," means the point $(q, p) = (12872, 4.50)$ lies on the graph as well.

We can use SymPy to find the linear function that passes through these two points, based on point-slope form:

In [3]:
slope = (4.50-5)/(12872 - 10100)
p = slope*(q - 10100) + 5
print(p)

6.82178932178932 - 0.00018037518037518*q


Now, we can also determine the revenue: the number of items sold times the price per item. We store the revenue $R(q)$ under the name R in Python.

In [4]:
R = p*q
print(R)

q*(6.82178932178932 - 0.00018037518037518*q)


### Maximizing Profit

Express the profit earned as a function of $q$. How many items should the company produce to maximize profit? What is the profit at that production level?

---

The profit is the revenue minus the cost. We can use the previously defined names R for revenue and C for cost to define profit in Python.

In [5]:
P = R - C
print(P)

q*(6.82178932178932 - 0.00018037518037518*q) - 2*q - 10000


To find the maximum profit, we first take the derivative of $P$ with SymPy's diff() function.

We can then find critical points by solving $P'(q) = 0$ with the solveset function. The list() function converts the results into a list so that we can take out values.

In [6]:
dP = diff(P, q)
crit_points = list(solveset(dP, q))
crit_points

[13366.0000000000]

We see that the maximum profit occurs when $13366$ items are sold. 

To find the profit at that production level, we subsitute this value for $q$ in $P(q)$ using the subs() function in SymPy.

In [7]:
P.subs(q, crit_points[0])

22224.0180375180

# Source

This example was adapted from an exercise in the following textbook.

Hughes-Hallett, D., Lock, P.F., Flath, D.E., Gleason, A.M., Lovelock, D., Quinney, D., Connally, E., Lozano, G.I., Rhea, K., Kalacioglu, S., McCallum, W.G., Spiegler, A.H., Lahme, B., Osgood, B.G., Tecosky-Feldman, J., Lomen, D.O., Patterson, C.L., Tucker, T.W., & Wootton, A.D. (2014). *Applied Calculus* (5th ed.). Wiley.